EclipseLink Snapshot 5.0.0.20240807 failing with JPQLException

We have identified a regression caused by the recent snapshot update. This regression specifically affects the more JPQL query, and it is currently blocking our ability to integrate the snapshot into Open Liberty. Consequently, we are unable to proceed with testing other critical fixes.
 An exception occurred while creating a query in EntityManager. Syntax error parsing [ FROM Prime WHERE (( numberId<=:max AND UPPER(romanNumeral) NOT LIKE '%VII' AND MOD(numberId,10)<>3 AND	numberId	>= :min)) ORDER BY even DESC, name]. 
Refer the logs:
```
Caused by: java.lang.IllegalArgumentException: An exception occurred while creating a query in EntityManager: 
Exception Description: Syntax error parsing [ FROM Prime WHERE (( numberId<=:max AND UPPER(romanNumeral) NOT LIKE '%VII' AND MOD(numberId,10)<>3 AND	numberId	>= :min)) ORDER BY even DESC, name]. 
[58, 70] The encapsulated expression is not a valid expression. (SELECT this FROM Prime WHERE ((numberId <= :max AND UPPER( [ UPPER(romanNumeral) ] ...
[96, 104] The first argument is not a valid expression. (SELECT this FROM Prime WHERE ((numberId <= :max AND UPPER(romanNumeral) NOT LIKE '%VII' AND MOD( [ MOD(numberId,10) ] ...
	at org.eclipse.persistence.internal.jpa.EntityManagerImpl.createQuery(EntityManagerImpl.java:1848)
	at org.eclipse.persistence.internal.jpa.EntityManagerImpl.createQuery(EntityManagerImpl.java:1869)
	at io.openliberty.data.internal.persistence.CursoredPageImpl.<init>(CursoredPageImpl.java:80)
Caused by: Exception [EclipseLink-0] (Eclipse Persistence Services - 5.0.0.v202408071314-43356e84b79e71022b1656a5462b0a72d70787a4): org.eclipse.persistence.exceptions.JPQLException
Exception Description: Syntax error parsing [ FROM Prime WHERE (( numberId<=:max AND UPPER(romanNumeral) NOT LIKE '%VII' AND MOD(numberId,10)<>3 AND	numberId	>= :min)) ORDER BY even DESC, name]. 
[58, 70] The encapsulated expression is not a valid expression. (SELECT this FROM Prime WHERE ((numberId <= :max AND UPPER( [ UPPER(romanNumeral) ] ...
[96, 104] The first argument is not a valid expression. (SELECT this FROM Prime WHERE ((numberId <= :max AND UPPER(romanNumeral) NOT LIKE '%VII' AND MOD( [ MOD(numberId,10) ] ...
	at org.eclipse.persistence.internal.jpa.jpql.HermesParser.buildException(HermesParser.java:175)
	at org.eclipse.persistence.internal.jpa.jpql.HermesParser.validate(HermesParser.java:351)
	at org.eclipse.persistence.internal.jpa.jpql.HermesParser.populateQueryImp(HermesParser.java:298)
	at org.eclipse.persistence.internal.jpa.jpql.HermesParser.buildQuery(HermesParser.java:180)
	at org.eclipse.persistence.internal.jpa.EJBQueryImpl.buildEJBQLDatabaseQuery(EJBQueryImpl.java:144)
	at org.eclipse.persistence.internal.jpa.EJBQueryImpl.buildEJBQLDatabaseQuery(EJBQueryImpl.java:120)
	at org.eclipse.persistence.internal.jpa.EJBQueryImpl.<init>(EJBQueryImpl.java:107)
	at org.eclipse.persistence.internal.jpa.EJBQueryImpl.<init>(EJBQueryImpl.java:91)
	at org.eclipse.persistence.internal.jpa.EntityManagerImpl.createQuery(EntityManagerImpl.java:1846)

	at componenttest.topology.utils.FATServletClient.assertTestResponse(FATServletClient.java:106)
	at componenttest.topology.utils.FATServletClient.runTest(FATServletClient.java:91)
	at componenttest.custom.junit.runner.SyntheticServletTest.invokeExplosively(SyntheticServletTest.java:49)
	at componenttest.custom.junit.runner.FATRunner$1.evaluate(FATRunner.java:204)
	at org.testcontainers.containers.FailureDetectingExternalResource$1.evaluate(FailureDetectingExternalResource.java:29)
	at componenttest.custom.junit.runner.FATRunner$2.evaluate(FATRunner.java:364)
	at componenttest.custom.junit.runner.FATRunner.run(FATRunner.java:178)
```

--- a/jpa/eclipselink.jpa.testapps/jpa.test.jpql/src/test/java/org/eclipse/persistence/testing/tests/jpa/jpql/advanced/JUnitJPQLJakartaDataNoAliasTest.java
+++ b/jpa/eclipselink.jpa.testapps/jpa.test.jpql/src/test/java/org/eclipse/persistence/testing/tests/jpa/jpql/advanced/JUnitJPQLJakartaDataNoAliasTest.java
@@ -100,6 +100,9 @@ public class JUnitJPQLJakartaDataNoAliasTest extends JUnitTestCase {
         suite.addTest(new JUnitJPQLJakartaDataNoAliasTest("testNoAliasFromWhere"));
         suite.addTest(new JUnitJPQLJakartaDataNoAliasTest("testNoAliasFromWhereAnd"));
         suite.addTest(new JUnitJPQLJakartaDataNoAliasTest("testNoAliasFromWhereAndUPPER"));
+        suite.addTest(new JUnitJPQLJakartaDataNoAliasTest("testNoAliasWhereWithBraces01"));
+        suite.addTest(new JUnitJPQLJakartaDataNoAliasTest("testNoAliasWhereWithBraces02"));
+        suite.addTest(new JUnitJPQLJakartaDataNoAliasTest("testNoAliasWhereWithBraces03"));
         suite.addTest(new JUnitJPQLJakartaDataNoAliasTest("testGeneratedSelectNoAliasFromWhere"));
         suite.addTest(new JUnitJPQLJakartaDataNoAliasTest("testGeneratedSelect"));
         suite.addTest(new JUnitJPQLJakartaDataNoAliasTest("testUpdateQueryLengthInAssignmentAndExpression"));
@@ -238,6 +241,50 @@ public class JUnitJPQLJakartaDataNoAliasTest extends JUnitTestCase {
         WrapperTypes tlWrapperTypes = (WrapperTypes) getPersistenceUnitServerSession().executeQuery(tlQuery);
         Assert.assertTrue("NoAliasFromWhereAndUPPER Test Failed", comparer.compareObjects(wrapperTypes, tlWrapperTypes));
     }
+
+    public void testNoAliasWhereWithBraces01() {
+        EntityManager em = createEntityManager();
+
+        Query wrapperTypesQuery = em.createQuery("FROM WrapperTypes WHERE (id = :idParam)");
+        wrapperTypesQuery.setParameter("idParam", wrapperId);
+        WrapperTypes wrapperTypes = (WrapperTypes) wrapperTypesQuery.getResultList().get(0);
+        clearCache();
+        ReadObjectQuery tlQuery = new ReadObjectQuery(WrapperTypes.class);
+        tlQuery.setSelectionCriteria(tlQuery.getExpressionBuilder().get("id").equal(wrapperId));
+
+        WrapperTypes tlWrapperTypes = (WrapperTypes) getPersistenceUnitServerSession().executeQuery(tlQuery);
+        Assert.assertTrue("NoAliasWhereWithBraces01 Test Failed", comparer.compareObjects(wrapperTypes, tlWrapperTypes));
+    }
+
+    public void testNoAliasWhereWithBraces02() {
+        EntityManager em = createEntityManager();
+
+        Query wrapperTypesQuery = em.createQuery("FROM WrapperTypes WHERE ((id = :idParam))");
+        wrapperTypesQuery.setParameter("idParam", wrapperId);
+        WrapperTypes wrapperTypes = (WrapperTypes) wrapperTypesQuery.getResultList().get(0);
+        clearCache();
+        ReadObjectQuery tlQuery = new ReadObjectQuery(WrapperTypes.class);
+        tlQuery.setSelectionCriteria(tlQuery.getExpressionBuilder().get("id").equal(wrapperId));
+
+        WrapperTypes tlWrapperTypes = (WrapperTypes) getPersistenceUnitServerSession().executeQuery(tlQuery);
+        Assert.assertTrue("NoAliasWhereWithBraces02 Test Failed", comparer.compareObjects(wrapperTypes, tlWrapperTypes));
+    }
+
+    public void testNoAliasWhereWithBraces03() {
+        EntityManager em = createEntityManager();
+
+        Query wrapperTypesQuery = em.createQuery("FROM WrapperTypes WHERE (id <= :idParam AND UPPER(stringData) NOT LIKE UPPER(:stringParam) AND UPPER(stringData) NOT LIKE 'dgdgs') ORDER BY id DESC, stringData");
+        wrapperTypesQuery.setParameter("idParam", wrapperId);
+        wrapperTypesQuery.setParameter("stringParam", "asdadads");
+        WrapperTypes wrapperTypes = (WrapperTypes) wrapperTypesQuery.getResultList().get(0);
+        clearCache();
+        ReadObjectQuery tlQuery = new ReadObjectQuery(WrapperTypes.class);
+        tlQuery.setSelectionCriteria(tlQuery.getExpressionBuilder().get("id").equal(wrapperId));
+
+        WrapperTypes tlWrapperTypes = (WrapperTypes) getPersistenceUnitServerSession().executeQuery(tlQuery);
+        Assert.assertTrue("NoAliasWhereWithBraces03 Test Failed", comparer.compareObjects(wrapperTypes, tlWrapperTypes));
+    }
+
 
     public void testGeneratedSelect() {
         EntityManager em = createEntityManager();

--- a/jpa/org.eclipse.persistence.jpa.jpql/src/main/java/org/eclipse/persistence/jpa/jpql/parser/AbstractExpression.java
+++ b/jpa/org.eclipse.persistence.jpa.jpql/src/main/java/org/eclipse/persistence/jpa/jpql/parser/AbstractExpression.java
@@ -525,7 +525,7 @@ public abstract class AbstractExpression implements Expression {
      * @return  Parent expression
      */
     public final ParentExpression getParentExpression() {
-        if (this instanceof ParentExpression parentExpression) {
+        if (!(this instanceof SubExpression) && this instanceof ParentExpression parentExpression) {
             return parentExpression;
         } else if (parent == null) {
             return null;

--- a/jpa/org.eclipse.persistence.jpa.jpql/src/main/java/org/eclipse/persistence/jpa/jpql/parser/AbstractExpression.java
+++ b/jpa/org.eclipse.persistence.jpa.jpql/src/main/java/org/eclipse/persistence/jpa/jpql/parser/AbstractExpression.java
@@ -525,10 +525,8 @@ public abstract class AbstractExpression implements Expression {
      * @return  Parent expression
      */
     public final ParentExpression getParentExpression() {
-        if (!(this instanceof SubExpression) && this instanceof ParentExpression parentExpression) {
-            return parentExpression;
-        } else if (parent == null) {
-            return null;
+        if (!(this.isSubExpression()) && (this.isParentExpression())) {
+            return (ParentExpression)this;
         } else {
             return parent.getParentExpression();
         }
@@ -604,6 +602,10 @@ public abstract class AbstractExpression implements Expression {
         return false;
     }
 
+    public boolean isParentExpression() {
+        return false;
+    }
+
     /**
      * Determines whether the parsing is complete based on what is left in the given text. The text
      * is never empty.
@@ -640,6 +642,10 @@ public abstract class AbstractExpression implements Expression {
      */
     protected boolean isTolerant() {
         return getRoot().isTolerant();
+    }
+
+    protected boolean isSubExpression() {
+        return false;
     }
 
     /**

--- a/jpa/org.eclipse.persistence.jpa.jpql/src/main/java/org/eclipse/persistence/jpa/jpql/parser/IdentificationVariable.java
+++ b/jpa/org.eclipse.persistence.jpa.jpql/src/main/java/org/eclipse/persistence/jpa/jpql/parser/IdentificationVariable.java
@@ -95,8 +95,24 @@ public final class IdentificationVariable extends AbstractExpression {
      */
     public IdentificationVariable(AbstractExpression parent, String identificationVariable) {
         super(parent, identificationVariable);
-        if (!Expression.THIS.equalsIgnoreCase(identificationVariable) && getParentExpression().isGenerateThisPrefix()) {
+        //In subqueries "this" generation is not allowed. There are expected qualified IdentificationVariable from query string
+        if (!Expression.THIS.equalsIgnoreCase(identificationVariable) && getParentExpression().isGenerateThisPrefix() && !isInsideSubquery()) {
             this.setVirtualIdentificationVariable(Expression.THIS);
+        }
+    }
+
+    private boolean isInsideSubquery() {
+        boolean result = isSubquery(this);
+        return result;
+    }
+
+    private boolean isSubquery(Expression expression) {
+        if (expression == null) {
+            return false;
+        } else if (expression instanceof AbstractSelectStatement && expression.getParent() != null && expression.getParent() instanceof SubExpression) {
+            return true;
+        } else {
+            return isSubquery(expression.getParent());
         }
     }
 

--- a/jpa/org.eclipse.persistence.jpa.jpql/src/main/java/org/eclipse/persistence/jpa/jpql/parser/JPQLExpression.java
+++ b/jpa/org.eclipse.persistence.jpa.jpql/src/main/java/org/eclipse/persistence/jpa/jpql/parser/JPQLExpression.java
@@ -266,6 +266,11 @@ public final class JPQLExpression extends AbstractExpression implements ParentEx
         this.generateThisPrefix = generateThisPrefix;
     }
 
+    @Override
+    public boolean isParentExpression() {
+        return true;
+    }
+
     public boolean isJakartaData() {
         return this.jakartaData;
     }

--- a/jpa/org.eclipse.persistence.jpa.jpql/src/main/java/org/eclipse/persistence/jpa/jpql/parser/ParentExpression.java
+++ b/jpa/org.eclipse.persistence.jpa.jpql/src/main/java/org/eclipse/persistence/jpa/jpql/parser/ParentExpression.java
@@ -23,4 +23,5 @@ public interface ParentExpression extends Expression {
 
     void setGenerateThisPrefix(boolean generateThisPrefix);
 
+    boolean isParentExpression();
 }

--- a/jpa/org.eclipse.persistence.jpa.jpql/src/main/java/org/eclipse/persistence/jpa/jpql/parser/SubExpression.java
+++ b/jpa/org.eclipse.persistence.jpa.jpql/src/main/java/org/eclipse/persistence/jpa/jpql/parser/SubExpression.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2006, 2021 Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2006, 2024 Oracle and/or its affiliates. All rights reserved.
  * Copyright (c) 2024 Contributors to the Eclipse Foundation. All rights reserved.
  *
  * This program and the accompanying materials are made available under the
@@ -55,6 +55,16 @@ public final class SubExpression extends AbstractSingleEncapsulatedExpression im
     @Override
     public void setGenerateThisPrefix(boolean generateThisPrefix) {
         this.generateThisPrefix = generateThisPrefix;
+    }
+
+    @Override
+    public boolean isParentExpression() {
+        return true;
+    }
+
+    @Override
+    public boolean isSubExpression() {
+        return true;
     }
 
     @Override

--- a/jpa/org.eclipse.persistence.jpa.jpql/src/test/java/org/eclipse/persistence/jpa/tests/jpql/parser/JPQLExpressionTestJakartaData.java
+++ b/jpa/org.eclipse.persistence.jpa.jpql/src/test/java/org/eclipse/persistence/jpa/tests/jpql/parser/JPQLExpressionTestJakartaData.java
@@ -87,8 +87,26 @@ public final class JPQLExpressionTestJakartaData extends JPQLParserTest {
     }
 
     @Test
-    public void testNoAliasFromWhere() {
+    public void testNoAliasFromWhere01() {
         JPQLExpression jpqlExpression = checkAliasFrom("SELECT this FROM NoAliasEntity WHERE id1 = :id1", Expression.THIS);
+        checkAliasesWhere(jpqlExpression, Expression.THIS);
+    }
+
+    @Test
+    public void testNoAliasFromWhere02() {
+        JPQLExpression jpqlExpression = checkAliasFrom("SELECT this FROM NoAliasEntity WHERE (id1 = :id1)", Expression.THIS);
+        checkAliasesWhere(jpqlExpression, Expression.THIS);
+    }
+
+    @Test
+    public void testNoAliasFromWhere03() {
+        JPQLExpression jpqlExpression = checkAliasFrom("SELECT this FROM NoAliasEntity WHERE (ID <= :idParam AND UPPER(name) NOT LIKE UPPER(:nameParam) AND UPPER(name) NOT LIKE 'dgdgs') ORDER BY ID DESC, name", Expression.THIS);
+        checkAliasesWhere(jpqlExpression, Expression.THIS);
+    }
+
+    @Test
+    public void testNoAliasFromWhere04() {
+        JPQLExpression jpqlExpression = checkAliasFrom("SELECT this FROM NoAliasEntity WHERE (:rate * ID <= :max AND :rate * ID >= :min) ORDER BY name", Expression.THIS);
         checkAliasesWhere(jpqlExpression, Expression.THIS);
     }
 

